domains/certificates/Query.API/API/DataSyncSyncer/DataSyncSyncerWorker.cs

```diff
@@ -13,41 +17,94 @@ namespace API.DataSyncSyncer;
 internal class DataSyncSyncerWorker : BackgroundService
 {
     private readonly ILogger<DataSyncSyncerWorker> logger;
-    private readonly IEventStore eventStore;
-    private readonly string? gsrn;
+    private readonly IIntegrationEventBus integrationEventBus;
+    private readonly IDataSync dataSync;
+    private readonly List<MasterData> masterData;
+    private readonly Dictionary<string, DateTimeOffset> periodStartTimeDictionary;
 
-    public DataSyncSyncerWorker(ILogger<DataSyncSyncerWorker> logger, IEventStore eventStore,
-        MockMasterDataCollection collection)
+    public DataSyncSyncerWorker(
+        ILogger<DataSyncSyncerWorker> logger,
+        MockMasterDataCollection collection,
+        IIntegrationEventBus queue,
+        IDataSync dataSync
+    )
     {
         this.logger = logger;
-        this.eventStore = eventStore;
-        var masterData = collection.Data.FirstOrDefault();
-        gsrn = masterData?.GSRN ?? null;
+        masterData = collection.Data.ToList();
+        integrationEventBus = queue;
+        this.dataSync = dataSync;
+        periodStartTimeDictionary = masterData
+            .Where(it => !string.IsNullOrWhiteSpace(it.GSRN))
+            .ToDictionary(m => m.GSRN, m => m.MeteringPointOnboardedStartDate);
     }
 
     protected override async Task ExecuteAsync(CancellationToken stoppingToken)
     {
-        if (string.IsNullOrWhiteSpace(gsrn))
+        while (!stoppingToken.IsCancellationRequested)
         {
-            logger.LogWarning("No master data loaded. Will not produce any events");
+            await SleepToNearestHour(stoppingToken);
+
+            foreach (var data in masterData)
+            {
+                var measurements = await FetchMeasurements(data.GSRN, data.MeteringPointOwner);
```

Comment:
The stoppingToken should be passed here to ensure that calls to DataSync will be cancelled as fast as possible in case the stoppingToken signals a cancellation.